pkg/workceptor/workceptor.go

```diff
@@ -678,25 +688,91 @@ func (w *Workceptor) unitStatusForCFR(unitID string) (map[string]interface{}, er
 	return retMap, nil
 }
 
+func (w *Workceptor) attemptRemoteAction(remoteNodeID, remoteUnitID, remoteAction string, unit *workUnit) error {
+	timeout := time.After(300 * time.Second)
+	retryInterval := 500 * time.Millisecond
+	shouldTimeout := func() bool {
+		select {
+		case <-timeout:
+			logger.Error("%s of remote work unit timed out", remoteAction)
+			return true
+		default:
+			time.Sleep(retryInterval)
+			return false
+		}
+	}
+
+	for {
+		conn, _, err := w.connectToRemote(remoteNodeID)
+		if err != nil {
+			logger.Error("Connection failed to %s: %s\n", remoteNodeID, err)
+		}
+
+		if conn != nil {
+			_, err = conn.Write([]byte(fmt.Sprintf("work %s %s\n", remoteAction, remoteUnitID)))
+			if err != nil {
+				logger.Error("Write error sending to %s: %s\n", remoteNodeID, err)
+			} else {
+				conn.Close()
+				break
+			}
+			conn.Close()
+		}
+
+		if shouldTimeout() {
+			return nil
+		}
+	}
+
+	if remoteAction == "cancel" {
+		var state int
+		for {
+			w.activeUnitsLock.RLock()
+			state = unit.status.State
+			w.activeUnitsLock.RUnlock()
+			if state == WorkStateSucceeded || state == WorkStateFailed {
+				return nil
+			}
+
+			if shouldTimeout() {
+				return nil
+			}
+		}
+	}
+	return nil
+}
+
 // CancelUnit cancels a unit, killing it if it is still running
 func (w *Workceptor) CancelUnit(unitID string) error {
 	//TODO: Handle remote cancellation
 	w.scanForUnits()
 	w.activeUnitsLock.RLock()
+	defer w.activeUnitsLock.RUnlock()
```

Comment:
Golang can't stack locks, so if you're holding this lock all the way to the end of the function, you can't call any _other_ function that tries to acquire the lock.  I guess the original reason I did this was for the cancel call, which is now moved to a goroutine.  Are we sure the `unit.status.Save` call is okay?